tensorflow/core/kernels/data/dataset_test_base.cc

```diff
@@ -560,5 +560,203 @@ Status DatasetOpsTestBase::AddDatasetInput(
+    const DatasetBase& dataset, int64 expected_cardinality) {
+  if (dataset.Cardinality() != expected_cardinality) {
+    return errors::Internal(
+        "The test for Dataset::Cardinality() failed: expected ",
+        expected_cardinality, " but got ", dataset.Cardinality(), ".");
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateDatasetSave(const DatasetBase& dataset) {
+  std::unique_ptr<SerializationContext> serialization_context;
+  TF_RETURN_IF_ERROR(CreateSerializationContext(&serialization_context));
+  VariantTensorData data;
+  VariantTensorDataWriter writer(&data);
+  Status status = dataset.Save(serialization_context.get(), &writer);
+  if (!status.ok()) {
+    return errors::Internal("The test for Dataset::Save() failed: ",
+                            status.ToString());
+  }
+  TF_RETURN_IF_ERROR(writer.Flush());
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateDatasetIsStateful(const DatasetBase& dataset,
+                                                     bool expected_stateful) {
+  if (dataset.IsStateful() != expected_stateful) {
+    return errors::Internal(
+        "The test for Dataset::IsStateful() failed: expected ",
+        expected_stateful, " but got ", dataset.IsStateful(), ".");
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateIteratorOutputDtypes(
+    const IteratorBase& iterator,
+    const DataTypeVector& expected_output_dtypes) {
+  Status status =
+      VerifyTypesMatch(iterator.output_dtypes(), expected_output_dtypes);
+  if (!status.ok()) {
+    return errors::Internal("The test for Iterator::output_dtypes() failed: ",
+                            status.ToString());
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateIteratorOutputShapes(
+    const IteratorBase& iterator,
+    const std::vector<PartialTensorShape>& expected_output_shapes) {
+  Status status =
+      VerifyShapesCompatible(iterator.output_shapes(), expected_output_shapes);
+  if (!status.ok()) {
+    return errors::Internal("The test for Iterator::output_shapes() failed: ",
+                            status.ToString());
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateIteratorPrefix(
+    const IteratorBase& iterator, const string& expected_iterator_prefix) {
+  if (iterator.prefix() != expected_iterator_prefix) {
+    return errors::Internal("The test for Iterator::prefix() failed: expected ",
+                            expected_iterator_prefix, " but got ",
+                            iterator.prefix());
+  }
+  return Status::OK();
+}
+
+Status DatasetOpsTestBase::EvaluateIteratorGetNext(
+    IteratorBase* iterator, IteratorContext* iterator_context,
+    const std::vector<Tensor>& expected_outputs, bool compare_order) {
+  bool end_of_sequence = false;
+  std::vector<Tensor> out_tensors;
+  while (!end_of_sequence) {
+    std::vector<Tensor> next;
+    TF_EXPECT_OK(iterator->GetNext(iterator_context, &next, &end_of_sequence));
```

Comment:
should this be `TF_RETURN_IF_ERROR` instead of `TF_EXPECT_OK`?

Reply:
Yes, changed it to be `TF_RETURN_IF_ERROR`.